### ISSUE TITLE
Use `libreadline-dev` instead of `libreadline6-dev` on Debian (≥stretch)

### ISSUE THREAD

**Issue by ngollan** [rvm/rvm#3824]
## Description

Debian testing does not have a `libreadline6-dev` package anymore, but a "generic" `libreadline-dev` that is currently at version 7.0. The stable rvm release thus fails to build rubies unless one tells autolibs to kindly shut up.

## Steps to reproduce

1. Install Debian testing
2. Install rvm (stable)
3. rvm install _\<any Ruby\>_ → fails due to a missing `libreadline6-dev` package

## Expected behavior

Install Ruby using installed `libreadline` headers and libraries.

## Actual behavior

For demonstration purposes result of trying to install Ruby 2.1.4, result is the same for all other versions up to 2.3.3.
```
$ rvm --version
rvm 1.27.0 (latest) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]
$ rvm install 2.1.4
Searching for binary rubies, this might take some time.
No binary rubies available for: debian/stretch_sid/x86_64/ruby-2.1.4.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Missing required packages: libreadline6-dev
RVM autolibs is now configured with mode '2' =>
  'Allow RVM to use package manager if found, fail if dependencies are missing. This is default.',
please run `rvm autolibs enable` to let RVM do its job or run and read `rvm autolibs [help]`
or visit https://rvm.io/rvm/autolibs for more information.
Requirements installation failed with status: 1.
```

**Comment by pkuczynski** [comment]
How can we recognize if RVM is running in Debian testing? Can you post `rvm info` output please?

**Comment by ngollan** [comment]
```ruby-2.3.3:

  system:
    uname:       "Linux metal 4.8.0-1-amd64 #1 SMP Debian 4.8.5-1 (2016-10-28) x86_64 GNU/Linux"
    system:      "debian/stretch_sid/x86_64"
    bash:        "/bin/bash => GNU bash, version 4.4.5(1)-release (x86_64-pc-linux-gnu)"
    zsh:         " => not installed"

  rvm:
    version:      "rvm 1.27.0 (latest) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]"
    updated:      "4 days 6 hours 59 minutes 43 seconds ago"
    path:         "/home/ngollan/.rvm"

  ruby:
    interpreter:  "ruby"
    version:      "2.3.3p222"
    date:         "2016-11-21"
    platform:     "x86_64-linux"
    patchlevel:   "2016-11-21 revision 56859"
    full_version: "ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-linux]"

  homes:
    gem:          "/home/ngollan/.rvm/gems/ruby-2.3.3"
    ruby:         "/home/ngollan/.rvm/rubies/ruby-2.3.3"

  binaries:
    ruby:         "/home/ngollan/.rvm/rubies/ruby-2.3.3/bin/ruby"
    irb:          "/home/ngollan/.rvm/rubies/ruby-2.3.3/bin/irb"
    gem:          "/home/ngollan/.rvm/rubies/ruby-2.3.3/bin/gem"
    rake:         "/home/ngollan/.rvm/rubies/ruby-2.3.3/bin/rake"

  environment:
    PATH:         "/home/ngollan/.rvm/gems/ruby-2.3.3/bin:/home/ngollan/.rvm/gems/ruby-2.3.3@global/bin:/home/ngollan/.rvm/rubies/ruby-2.3.3/bin:/home/ngollan/.rvm/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games"
    GEM_HOME:     "/home/ngollan/.rvm/gems/ruby-2.3.3"
    GEM_PATH:     "/home/ngollan/.rvm/gems/ruby-2.3.3:/home/ngollan/.rvm/gems/ruby-2.3.3@global"
    MY_RUBY_HOME: "/home/ngollan/.rvm/rubies/ruby-2.3.3"
    IRBRC:        "/home/ngollan/.rvm/rubies/ruby-2.3.3/.irbrc"
    RUBYOPT:      ""
    gemset:       ""
```

**Comment by pkuczynski** [comment]
I guess we could hook up on the `lsb_release:Codename` value. I would suggest to implement it in the following way (pseudo code):

```
case "$lsb_release:Codename" in
  (stretch)
    requirements_check libreadline-dev
    ;;
  (*)
    requirements_check libreadline6-dev
    ;;
esac
```

Would that work? 

Would be better if we could hookup on `Release` field >8, but until Stretch is released, this field has value `testing` - correct?

**Comment by ngollan** [comment]
Current stable has 8.6 in the "Release" field.

It would probably make sense to check for a major version >8 OR any non-numeric value like "testing" or "unstable" to stay safe once the next stable is released, which could be any century now.

**Comment by pkuczynski** [comment]
I prepared a patch. Could you test it and let me know if it works?

```
rvm get branch /bugfix/3824-libreadline
rvm install 2.1.4
```

**Comment by pkuczynski** [comment]
@ngollan have you had a chance to test it?

**Comment by ngollan** [comment]
I'll try to give it a spin today (don't quite want to fumble my office dev environment), if not on the weekend.

**Comment by ngollan** [comment]
I don't think this is working, now it fails to find just about every dependency including `libreadline6-dev`:

```
$ rvm --version
rvm 1.27.0 (bugfix/3824-libreadline) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]

$ rvm install --autolibs=read-fail 2.3.3
ruby-2.3.3 - #removing src/ruby-2.3.3..
Searching for binary rubies, this might take some time.
No binary rubies available for: debian/stretch_sid/x86_64/ruby-2.3.3.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Missing required packages: bash curl patch bzip2 gawk g++ gcc make libc6-dev patch openssl libreadline6 curl zlib1g zlib1g-dev libssl-dev libyaml-dev libsqlite3-dev sqlite3 libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline6-dev
RVM autolibs is now configured with mode '2' =>
  'Allow RVM to use package manager if found, fail if dependencies are missing. This is default.',
please run `rvm autolibs enable` to let RVM do its job or run and read `rvm autolibs [help]`
or visit https://rvm.io/rvm/autolibs for more information.
Requirements installation failed with status: 1.
```

**Comment by pkuczynski** [comment]
I reverted one change which might cause the issue. Could you please retry again:

```
rvm get branch /bugfix/3824-libreadline
rvm install 2.1.4
```

**Comment by ngollan** [comment]
This restores the behaviour of current stable in that it requires `libreadline6-dev` on Debian testing.

**Comment by pkuczynski** [comment]
I see. I will try to come up with a patch this week. I have an idea what might be the source of the problem. Thank you for you support and testing!

**Comment by tsolar** [comment]
I have been investigating, and if current Debian installation is not stable, it does not have a number in lsb_release 'Release' field. Codename field is kind of more reliable. Codename uses the "Toy Story" codename. I use "unstable" and it says "sid". So, you should check if Codename is "sid", or something like if "Release" has not any number then use `libreadline-dev`, otherwise if major version (extracting the "8" from "8.6", for example) is greater than 8 then use `libreadline-dev`, else use `libreadline6-dev`. 

Hope that helps!

**Comment by pkuczynski** [comment]
I know, and the code I put in the branch is already OK, but there is one small issue which I need to trace and fix. Thanks for the comment anyway...

**Comment by tsolar** [comment]
I can confirm it is working for me, using Debian sid (stretch/sid). I was having issues with another problem and it was confusing me. Thanks for the quick answer!

**Comment by pkuczynski** [comment]
@tsolar so this branch works well for you for Debian testing (stretch)?

**Comment by tsolar** [comment]
I am using Debian Sid.
```
$ cat /etc/debian_version 
stretch/sid
```

It is working for me
```
$ rvm --version
rvm 1.27.0 (bugfix/3824-libreadline) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]
```

I got these messages when I install a ruby version:

```
$ rvm install 2.3.1
Searching for binary rubies, this might take some time.
No binary rubies available for: debian/stretch_sid/x86_64/ruby-2.3.1.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Missing required packages: libssl-dev libreadline6-dev
Requirements installation successful.
Installing Ruby from source to: /home/tom/.rvm/rubies/ruby-2.3.1, this may take a while depending on your cpu(s)...
ruby-2.3.1 - #downloading ruby-2.3.1, this may take a while depending on your connection...
ruby-2.3.1 - #extracting ruby-2.3.1 to /home/tom/.rvm/src/ruby-2.3.1....
ruby-2.3.1 - #configuring..........................................................
ruby-2.3.1 - #post-configuration..
ruby-2.3.1 - #compiling..................................................................................
ruby-2.3.1 - #installing................
ruby-2.3.1 - #making binaries executable..
Installed rubygems 2.5.1 is newer than 2.4.8 provided with installed ruby, skipping installation, use --force to force installation.
ruby-2.3.1 - #gemset created /home/tom/.rvm/gems/ruby-2.3.1@global
ruby-2.3.1 - #importing gemset /home/tom/.rvm/gemsets/global.gems...............................................
ruby-2.3.1 - #generating global wrappers........
ruby-2.3.1 - #gemset created /home/tom/.rvm/gems/ruby-2.3.1
ruby-2.3.1 - #importing gemsetfile /home/tom/.rvm/gemsets/default.gems evaluated to empty gem list
ruby-2.3.1 - #generating default wrappers........
ruby-2.3.1 - #adjusting #shebangs for (gem irb erb ri rdoc testrb rake).
Install of ruby-2.3.1 - #complete 
Ruby was built without documentation, to build it run: rvm docs generate-ri
```

Notice the `Missing required packages: libssl-dev libreadline6-dev` line. Is it ok or expected?

**Comment by tsolar** [comment]
I realized I was using `rvm autolibs 1`.

I run `rvm autolibs reset`, and then `rvm install 2.3.3` and I got [this](https://gist.github.com/tsolar/23c73e4ede48569a570bb1bd64662141) :(

**Comment by kfrz** [comment]
@tsolar 

```
› rvm install ruby-2.3.1                 
Warning, new version of rvm available '1.28.0', you are using older version '1.27.0'.
You can disable this warning with:    echo rvm_autoupdate_flag=0 >> ~/.rvmrc
You can enable  auto-update  with:    echo rvm_autoupdate_flag=2 >> ~/.rvmrc
ruby-2.3.1 - #removing src/ruby-2.3.1 - please wait
Searching for binary rubies, this might take some time.
No binary rubies available for: debian/stretch_sid/x86_64/ruby-2.3.1.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Installing requirements for debian.
Updating system - please wait
kf password required for 'apt-get --quiet --yes update': 
Installing required packages: bash, curl, patch, bzip2, gawk, g++, gcc, make, libc6-dev, patch, openssl, libreadline6, curl, zlib1g, zlib1g-dev, libssl-dev, libyaml-dev, libsqlite3-dev, sqlite3, libgdbm-dev, libncurses5-dev, bison, pkg-config, libffi-dev, libgmp-dev, libreadline6-dev - please wait
Requirements installation successful.
Installing Ruby from source to: /home/kf/.rvm/rubies/ruby-2.3.1, this may take a while depending on your cpu(s)...
ruby-2.3.1 - #downloading ruby-2.3.1, this may take a while depending on your connection...
ruby-2.3.1 - #extracting ruby-2.3.1 to /home/kf/.rvm/src/ruby-2.3.1 - please wait
ruby-2.3.1 - #configuring - please wait
ruby-2.3.1 - #post-configuration - please wait
ruby-2.3.1 - #compiling - please wait
Error running '__rvm_make -j4',
showing last 15 lines of /home/kf/.rvm/log/1482787760_ruby-2.3.1/make.log
make[1]: *** [ext/openssl/all] Error 2
make[1]: *** Waiting for unfinished jobs....
installing default nkf libraries
compiling date_strptime.c
linking shared-object objspace.so
installing default date_core libraries
make[2]: Leaving directory '/home/kf/.rvm/src/ruby-2.3.1/ext/objspace'
linking shared-object date_core.so
make[2]: Leaving directory '/home/kf/.rvm/src/ruby-2.3.1/ext/date'
linking shared-object nkf.so
make[2]: Leaving directory '/home/kf/.rvm/src/ruby-2.3.1/ext/nkf'
make[1]: Leaving directory '/home/kf/.rvm/src/ruby-2.3.1'
uncommon.mk:203: recipe for target 'build-ext' failed
make: *** [build-ext] Error 2
+__rvm_make:0> return 2
There has been an error while running make. Halting the installation.
```

```
› uname -a
Linux stretchbook 4.7.0-1-amd64 #1 SMP Debian 4.7.8-1 (2016-10-19) x86_64 GNU/Linux
```

```
› cat /etc/debian_version
stretch/sid
```

```
› rvm --version
rvm 1.28.0 (master) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io/]
```

Same thing happened on the 1.27 bugfix from above.

**Comment by raelgc** [comment]
Hi @kfrz. I can see your `rvm` is not updated. Can you update it and try again, please?

**Comment by pkuczynski** [comment]
This has not been yet completed and merged to master...

**Comment by kfrz** [comment]
@raelgc same thing when I use 1.28. I tried to compile the ruby manually as such:

```
./configure && make
```

The output of which looks fine until it begins to compile readline, at which point I see

```
make[2]: Entering directory '/home/kf/.rvm/src/ruby-2.3.1/ext/openssl'
compiling ossl_x509cert.c
In file included from ossl.h:57:0,
                 from ossl_x509cert.c:10:
/usr/include/openssl/asn1_mac.h:10:2: error: #error "This file is obsolete; please update your software."
 #error "This file is obsolete; please update your software."
  ^~~~~
In file included from ossl.h:221:0,
                 from ossl_x509cert.c:10:
openssl_missing.h:78:35: error: macro "EVP_MD_CTX_create" passed 1 arguments, but takes just 0
 EVP_MD_CTX *EVP_MD_CTX_create(void);
                                   ^
In file included from /usr/include/openssl/x509.h:23:0,
                 from /usr/include/openssl/x509v3.h:14,
                 from ossl.h:58,
                 from ossl_x509cert.c:10:
openssl_missing.h:82:6: error: expected declaration specifiers or ‘...’ before ‘(’ token
 void EVP_MD_CTX_init(EVP_MD_CTX *ctx);
      ^
openssl_missing.h:90:6: error: expected declaration specifiers or ‘...’ before ‘(’ token
 void EVP_MD_CTX_destroy(EVP_MD_CTX *ctx);
      ^
ossl_x509cert.c: In function ‘DupX509CertPtr’:
ossl_x509cert.c:125:5: warning: implicit declaration of function ‘CRYPTO_add’ [-Wimplicit-function-declaration]
     CRYPTO_add(&x509->references, 1, CRYPTO_LOCK_X509);
     ^~~~~~~~~~
ossl_x509cert.c:125:21: error: dereferencing pointer to incomplete type ‘X509 {aka struct x509_st}’
     CRYPTO_add(&x509->references, 1, CRYPTO_LOCK_X509);
                     ^~
ossl_x509cert.c:125:38: error: ‘CRYPTO_LOCK_X509’ undeclared (first use in this function)
     CRYPTO_add(&x509->references, 1, CRYPTO_LOCK_X509);
                                      ^~~~~~~~~~~~~~~~
ossl_x509cert.c:125:38: note: each undeclared identifier is reported only once for each function it appears in
Makefile:301: recipe for target 'ossl_x509cert.o' failed
make[2]: *** [ossl_x509cert.o] Error 1
make[2]: Leaving directory '/home/kf/.rvm/src/ruby-2.3.1/ext/openssl'
exts.mk:208: recipe for target 'ext/openssl/all' failed
make[1]: *** [ext/openssl/all] Error 2
make[1]: Leaving directory '/home/kf/.rvm/src/ruby-2.3.1'
uncommon.mk:203: recipe for target 'build-ext' failed
make: *** [build-ext] Error 2
```

I have set `rvm autolibs 3`

**Comment by tsolar** [comment]
A workaround i am using is:

* add stable repos to `/etc/apt/sources.list`, like 
```
deb http://ftp.us.debian.org/debian stable main
```
* install `libreadline6-dev`, `libreadline6` and `libssl-dev` from stable repo (resolve yourself any package conflicts ;) 
``` 
# apt install libreadline6-dev/stable libreadline6/stable libssl-dev/stable
```
* to prevent changes due to updates, mark as "hold" these packages
```
# apt-mark hold libreadline6
# apt-mark hold libreadline6-dev
# apt-mark hold libssl-dev
```

I was thinking about instead of check Debian version, check if packages are available with `apt-cache show foobar`? `apt-cache show foobar` returns a non-zero exit code if no packages matching `foobar` are found.

**Comment by cg2v** [comment]
ubuntu 16.10 is also affected

**Comment by eriicj** [comment]
Thanks **tsolar** ! Have the same problem, use your workaround, has no more problem ! :)

**Comment by pkuczynski** [comment]
@cg2v there is no `libssl1.0-dev` package for Ubuntu 16.10, but only `libssl-dev`. However there is `libssl1.0.0` package - not sure if that is equivalent to `libssl1.0-dev` or not?

**Comment by tsolar** [comment]
I tried `rvm get head && rvm reload && rvm autolibs reset && rvm reinstall 2.3.3` and I got this:

```
[...]
Searching for binary rubies, this might take some time.
No binary rubies available for: debian/9/x86_64/ruby-2.3.3.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Installing requirements for debian.
Updating system.....
Installing required packages: bash, curl, patch, bzip2, gawk, libssl1.0-dev, make, libc6-dev, patch, openssl, libreadline6, libreadline6-dev, curl, zlib1g, zlib1g-dev, libyaml-dev, libsqlite3-dev, sqlite3, libgmp-dev, libgdbm-dev, libncurses5-dev, bison, pkg-config, libffi-dev, libgmp-dev, libreadline-dev.......
Error running 'requirements_debian_libs_install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl libreadline6 libreadline6-dev curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev',
showing last 15 lines of /home/tom/.rvm/log/1486586347_ruby-2.3.3/package_install_bash_curl_patch_bzip2_gawk_libssl1.0-dev_make_libc6-dev_patch_openssl_libreadline6_libreadline6-dev_curl_zlib1g_zlib1g-dev_libyaml-dev_libsqlite3-dev_sqlite3_libgmp-dev_libgdbm-dev_libncurses5-.log
sqlite3 ya está en su versión más reciente (3.16.2-2).
zlib1g ya está en su versión más reciente (1:1.2.8.dfsg-5).
zlib1g-dev ya está en su versión más reciente (1:1.2.8.dfsg-5).
No se pudieron instalar algunos paquetes. Esto puede significar que
usted pidió una situación imposible o, si está usando la distribución
inestable, que algunos paquetes necesarios aún no se han creado o se
han sacado de «Incoming».
La siguiente información puede ayudar a resolver la situación:

Los siguientes paquetes tienen dependencias incumplidas:
 libreadline-dev : Entra en conflicto: libreadline6-dev
 libreadline6-dev : Depende: libreadline6 (= 6.3-8+b3) pero 6.3-9 va a ser instalado
E: No se pudieron corregir los problemas, usted ha retenido paquetes rotos.
++ return 100
++ return 100
Requirements installation failed with status: 100.
```

It seems it is still trying to install `libreadline6-dev`?

**Comment by pkuczynski** [comment]
Could you please run `rvm get master` and post gist with the output of the following commands:

```
cat /etc/debian_version
rvm info
rvm reinstall 2.3.3
```

I extended the `rvm info` to better understand where is the problem. Could be that we incorrectly recognize

**Comment by ngollan** [comment]
Building on debian testing (reporting as 9.0) now appears to be completely broken in master as well; it (still) expects both readline6-dev and readline-dev:

https://gist.github.com/ngollan/39fdb49e5ec31e21b1971a96e7214323

**Comment by pkuczynski** [comment]
Apologies. It was a bad merge on my side. Could you `rvm get master` and try again please?

**Comment by tsolar** [comment]
Oh no. Guess what. There is no package `libreadline` in Debian :/
There is `libreadline-dev` and `libreadline7`. 

`libreadline-dev` depends on `libreadline7` now, so I guess it is not needed to put it in the requirements.
Just putting `libreadline-dev` should work.

**Comment by ngollan** [comment]
For libssl development packages, there's a `libssl-dev` package which is currently at 1.1.0 versions for testing and unstable, and a separate `libssl1.0-dev`.

The 1.0 package is marked as transitional, so it doesn't sound like the best thing to target.

**Comment by tsolar** [comment]
@ngollan but ruby works with libssl1.0, not with libssl1.1.0, right? Maybe new versions of ruby (>2.4) can work with 1.1?

**Comment by ngollan** [comment]
@tsolar seems like it, I probably lost the 1.0 headers and libs during a dist-upgrade recently. Building 2.3.3 with 1.1 yields some eclectic errors.

**Comment by pkuczynski** [comment]
I just pushed a fix to master for using only `libreadline-dev`. When it comes to `libssl1.0` it is required for ruby older than 2.4, and is only required by the older versions. So all should work now. Let me know if it still does not?

**Comment by tsolar** [comment]
Yay! It worked!

**Comment by pkuczynski** [comment]
Finally! Sorry it took so much time...

**Comment by amontalban** [comment]
Hey guys,

I'm having issues with RVM on Debian 8 (Jessie), it works ok on stable 1.28.0 but with master branch it shows this error:

```
jenkins-slave@f188e92e4d32:~/workspace/SaltStack-MultiJob$ bash -c " source ~/.rvm/scripts/rvm && rvm use --install --create . && export > rvm.env"
RVM used your Gemfile for selecting Ruby, it is all fine - Heroku does that too,
you can ignore these warnings with 'rvm rvmrc warning ignore /home/jenkins-slave/workspace/SaltStack-MultiJob/Gemfile'.
To ignore the warning for all files run 'rvm rvmrc warning ignore allGemfiles'.

ruby-2.2.5 is not installed.
To install do: 'rvm install ruby-2.2.5'
RVM used your Gemfile for selecting Ruby, it is all fine - Heroku does that too,
you can ignore these warnings with 'rvm rvmrc warning ignore /home/jenkins-slave/workspace/SaltStack-MultiJob/Gemfile'.
To ignore the warning for all files run 'rvm rvmrc warning ignore allGemfiles'.

ruby-2.2.5 is not installed - installing.
No binary rubies available for: debian/8/x86_64/ruby-2.2.5.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Missing required packages: bash curl patch bzip2 gawk libssl-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadl
ine6-dev
RVM autolibs is now configured with mode '2' =>
  'Allow RVM to use package manager if found, fail if dependencies are missing. This is default.',
please run `rvm autolibs enable` to let RVM do its job or run and read `rvm autolibs [help]`
or visit https://rvm.io/rvm/autolibs for more information.
Requirements installation failed with status: 1.
ruby-2.2.5 is not installed - installing.
Searching for binary rubies, this might take some time.
No binary rubies available for: debian/8/x86_64/ruby-2.2.5.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Missing required packages: bash curl patch bzip2 gawk libssl-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadl
ine6-dev
RVM autolibs is now configured with mode '2' =>
  'Allow RVM to use package manager if found, fail if dependencies are missing. This is default.',
please run `rvm autolibs enable` to let RVM do its job or run and read `rvm autolibs [help]`
or visit https://rvm.io/rvm/autolibs for more information.
Requirements installation failed with status: 1.
Using /home/jenkins-slave/.rvm/gems/ruby-2.2.5
Warning! Executable 'ruby' missing, something went wrong with this ruby installation!
Warning! Executable 'gem' missing, something went wrong with this ruby installation!
Warning! Executable 'irb' missing, something went wrong with this ruby installation!
```

**rvm version:**

```
jenkins-slave@f188e92e4d32:~/workspace/SaltStack-MultiJob$ rvm version
RVM used your Gemfile for selecting Ruby, it is all fine - Heroku does that too,
you can ignore these warnings with 'rvm rvmrc warning ignore /home/jenkins-slave/workspace/SaltStack-MultiJob/Gemfile'.
To ignore the warning for all files run 'rvm rvmrc warning ignore allGemfiles'.

ruby-2.2.5 is not installed.
To install do: 'rvm install ruby-2.2.5'
rvm 1.28.0 (master) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io/]
```

**rvm info:**
```
jenkins-slave@f188e92e4d32:~/.rvm$ rvm version
rvm 1.28.0 (master) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io/]
jenkins-slave@f188e92e4d32:~/.rvm$ rvm info

system:

  system:
    uname:        "Linux f188e92e4d32 4.4.0-47-generic #68-Ubuntu SMP Wed Oct 26 19:39:52 UTC 2016 x86_64 GNU/Linux"
    name:         "Debian"
    version:      "8"
    architecture: "x86_64"
    bash:         "/bin/bash => GNU bash, version 4.3.30(1)-release (x86_64-pc-linux-gnu)"
    zsh:          " => not installed"
    remote path:  "debian/8/x86_64"

  rvm:
    version:      "rvm 1.28.0 (master) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io/]"
    updated:      "17 minutes 3 seconds ago
    path:         "/home/jenkins-slave/.rvm"
    autolibs:     "[read-fail] Allow RVM to use package manager if found, fail if dependencies are missing. This is default."

  homes:
    gem:          "not set"
    ruby:         "not set"

  binaries:
    ruby:         ""
    irb:          ""
    gem:          ""
    rake:         ""

  environment:
    PATH:         "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/jenkins-slave/.rvm/bin"
    GEM_HOME:     ""
    GEM_PATH:     ""
    MY_RUBY_HOME: ""
    IRBRC:        ""
    RUBYOPT:      ""
    gemset:       ""
```

Any help is highly appreciated!

Thanks

**Comment by pkuczynski** [comment]
In the future please report a separate issue instead of polluting old one! Could you please try `rvm autolibs enable` and then run whatever you tried to run. This should solve all packages dependencies.

**Comment by amontalban** [comment]
Hi @pkuczynski,

Sorry if I polluted this issue, but I saw that it was still active and was related to the issue I have.

`rvm autolibs enable` works but I think it's not the right solution because:

1) I have all required dependencies installed.
2) I'm running rvm in a non-sudo user inside Docker.
3) rvm is not installing missing packages when I run `rvm install ruby-2.2.5`

Please let me know if you want me to create another ticket or paste relevant information here.

Thanks!

**Comment by pkuczynski** [comment]
Please open a new issue and specify your working env and which packages are missing

**Comment by carolinesalib** [comment]
HI!

I'm using Ubuntu 16.04 and try to install ruby 2.2.3, but still trying to install the nonexistent libssl1.0-dev. :sob: 

```
Installing required packages: bash, curl, patch, bzip2, gawk, libssl1.0-dev, make, libc6-dev, patch, openssl, curl, zlib1g, zlib1g-dev, libyaml-dev, libsqlite3-dev, sqlite3, libgmp-dev, libgdbm-dev, libncurses5-dev, bison, pkg-config, libffi-dev, libgmp-dev, libreadline-dev....
Error running 'requirements_debian_libs_install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev',
showing last 15 lines of /home/caroline/.rvm/log/1487335582_ruby-2.2.3/package_install_bash_curl_patch_bzip2_gawk_libssl1.0-dev_make_libc6-dev_patch_openssl_curl_zlib1g_zlib1g-dev_libyaml-dev_libsqlite3-dev_sqlite3_libgmp-dev_libgdbm-dev_libncurses5-dev_bison_pkg-config_libf.log
++ [[ -d /usr/local/sbin ]]
++ [[ :/home/caroline/.rvm/gems/ruby-2.4.0/bin:/home/caroline/.rvm/gems/ruby-2.4.0@global/bin:/home/caroline/.rvm/rubies/ruby-2.4.0/bin:/home/caroline/.rvm/bin:/home/caroline/bin:/home/caroline/.local/bin:/home/caroline/bin:/home/caroline/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games: != *\:\/\u\s\r\/\l\o\c\a\l\/\s\b\i\n\:* ]]
++ [[ -n '' ]]
++ command_to_run=(__rvm_sudo -p "%p password required for '$*': " "${command_to_run[@]}")
++ __rvm_sudo -p '%p password required for '\''apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev'\'': ' apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev
++ command sudo -p '%p password required for '\''apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev'\'': ' apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev
++ sudo -p '%p password required for '\''apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev'\'': ' apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev
Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package libssl1.0-dev
E: Couldn't find any package by glob 'libssl1.0-dev'
E: Couldn't find any package by regex 'libssl1.0-dev'
++ return 100
++ return 100
```

**Comment by raelgc** [comment]
@carolinesalib Can you, please, try with the [Ubuntu dedicated package](https://github.com/rvm/ubuntu_rvm)?

**Comment by carolinesalib** [comment]
Hi, @raelgc! Thanks for try to help me.

I removed rvm using the command `rm -rf ~/.rvm` and then I installed again using ppa of the link you send to me, but the problem persists.

When I try to install ruby 2.2.4 it is ok, but 2.2.3 he searchs for the libsssl-1.0-dev.

```
No binary rubies available for: debian/9/x86_64/ruby-2.4.0.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Installing requirements for debian.
Updating system...
Installing required packages: bash, curl, patch, bzip2, gawk, g++, gcc, libssl-dev, make, libc6-dev, patch, openssl, curl, zlib1g, zlib1g-dev, libyaml-dev, libsqlite3-dev, sqlite3, libgmp-dev, libgdbm-dev, libncurses5-dev, bison, pkg-config, libffi-dev, libgmp-dev, libreadline-dev....
Requirements installation successful.
```

```
No binary rubies available for: debian/9/x86_64/ruby-2.2.3.
Continuing with compilation. Please read 'rvm help mount' to get more information on binary rubies.
Checking requirements for debian.
Installing requirements for debian.
Updating system...
Installing required packages: bash, curl, patch, bzip2, gawk, libssl1.0-dev, make, libc6-dev, patch, openssl, curl, zlib1g, zlib1g-dev, libyaml-dev, libsqlite3-dev, sqlite3, libgmp-dev, libgdbm-dev, libncurses5-dev, bison, pkg-config, libffi-dev, libgmp-dev, libreadline-dev..
Error running 'requirements_debian_libs_install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev',
showing last 15 lines of /home/caroline/.rvm/log/1487347952_ruby-2.2.3/package_install_bash_curl_patch_bzip2_gawk_libssl1.0-dev_make_libc6-dev_patch_openssl_curl_zlib1g_zlib1g-dev_libyaml-dev_libsqlite3-dev_sqlite3_libgmp-dev_libgdbm-dev_libncurses5-dev_bison_pkg-config_libf.log
command(25): requirements_debian_libs_install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev
++ __rvm_try_sudo apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev
++ typeset -a command_to_run
++ typeset sudo_path sbin_path missing_paths
++ command_to_run=("$@")
++ ((  UID == 0  ))
++ apt-get --no-install-recommends --yes install bash curl patch bzip2 gawk libssl1.0-dev make libc6-dev patch openssl curl zlib1g zlib1g-dev libyaml-dev libsqlite3-dev sqlite3 libgmp-dev libgdbm-dev libncurses5-dev bison pkg-config libffi-dev libgmp-dev libreadline-dev
Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package libssl1.0-dev
E: Couldn't find any package by glob 'libssl1.0-dev'
E: Couldn't find any package by regex 'libssl1.0-dev'
++ return 100
++ return 100
Requirements installation failed with status: 100.
```

**Comment by raelgc** [comment]
@carolinesalib It's weird because your system is reporting itself as Debian. We even have pre-compiled binaries for Ubuntu x64 for those 2 versions.
